lib/mayaUsd/fileio/primUpdaterManager.cpp

```diff
@@ -0,0 +1,1097 @@
+                      "pruning at that point.",
+                      srcPath.GetText());
+                  return false;
+              }
+              auto relativeSrcPath = srcPath.MakeRelativePath(SdfPath::AbsoluteRootPath());
+              auto dstPath = dstRootParentPath.AppendPath(relativeSrcPath);
+
+              // Report PushCopySpecs() failure.
+              if (!updater->PushCopySpecs(srcLayer, srcPath, dstLayer, dstPath)) {
+                  throw MayaUsd::TraversalFailure(std::string("PushCopySpecs() failed."), srcPath);
+              }
+
+              // Continue normal traversal without pruning.
+              return true;
+          };
+
+    if (!MayaUsd::traverseLayer(srcLayer, srcRootPath, pushCopySpecsFn)) {
+        return false;
+    }
+
+    // Push end is a separate traversal, not a second phase of the same
+    // traversal, because it is post-order: parents are traversed after
+    // children.  This allows for proper parent lifescope, if push end
+    // deletes the Maya node (which is the default behavior).
+    if (isCopy) {
+        return true;
+    }
+
+    // SdfLayer::TraversalFn does not return a status, so must report
+    // failure through an exception.
+    auto pushEndFn = [&context, srcLayer](const SdfPath& primSpecPath) {
+        // We can be called with a primSpec path that is not a prim path
+        // (e.g. a property path like "/A.xformOp:translate").  This is not an
+        // error, just a no-op.
+        if (!primSpecPath.IsPrimPath()) {
+            return;
+        }
+
+        auto updater = createUpdater(srcLayer, primSpecPath, context);
+        if (!updater) {
+            TF_WARN(
+                "Could not create a prim updater for path %s during PushEnd() traversal, pruning "
+                "at that point.",
+                primSpecPath.GetText());
+            return;
+        }
+
+        // Report PushEnd() failure.
+        if (!updater->PushEnd(context)) {
+            throw MayaUsd::TraversalFailure(std::string("PushEnd() failed."), primSpecPath);
+        }
+    };
+
+    // SdfLayer::Traverse does not return a status, so must report failure
+    // through an exception.
+    try {
+        srcLayer->Traverse(srcRootPath, pushEndFn);
+    } catch (const MayaUsd::TraversalFailure& e) {
+        TF_WARN(
+            "PushEnd() layer traversal failed for path %s: %s",
+            e.path().GetText(),
+            e.reason().c_str());
+        return false;
+    }
+
+    return true;
+}
+
+class PushPullScope
+{
+public:
+    PushPullScope(bool& controlingFlag)
+    {
+        if (!controlingFlag) {
+            controlingFlag = true;
+            _controlingFlag = &controlingFlag;
+        }
+    }
+    ~PushPullScope()
+    {
+        if (_controlingFlag) {
+            *_controlingFlag = false;
+        }
+    }
+
+private:
+    bool* _controlingFlag { nullptr };
+};
+
+MString pullRootName("__mayaUsd__");
```

Comment:
Let's put it at the top together with other constants. Update the naming convention as well.